Add a setting to increase visibility of the first object in "Hidden" mod

Vidalee opened the pull request:
Fix issue [#2043](https://github.com/ppy/osu/issues/2043)

Vidalee committed:
Added setting to not hide the first object in Hidden mod

Vidalee committed:
Why this line got deleted :thinking:

Vidalee committed:
Add license header

Vidalee committed:
Fix naming rule

Vidalee committed:
Chesterton's fence

Vidalee committed:
Review changes

Vidalee commented:
Thanks for he review ! Will fix this build error tomorrow.

peppy commented on osu.Game.Rulesets.Osu/Mods/OsuModHidden.cs:
```diff
@@ -83,5 +88,18 @@ protected void ApplyHiddenState(DrawableHitObject drawable, ArmedState state)
+        {
+            increaseFirstObjectVisibility = config.GetBindable<bool>(OsuSetting.IncreaseFirstObjectVisibility);
+            //This starts the process of applying the mod effects. We start it here since this is the last void called.
+            applyMod();
```
this shouldn't be called here. this call itself should be before other application calls.

peppy commented on osu.Game.Rulesets.Osu/Mods/OsuModHidden.cs:
```diff
@@ -7,32 +7,37 @@
 
         private const double fade_in_duration_multiplier = 0.4;
         private const double fade_out_duration_multiplier = 0.3;
+        private bool increaseFirstObjectVisibility = true;
```
store the bindable, not just a bool

peppy commented on osu.Game.Rulesets.Osu/Mods/OsuModHidden.cs:
```diff
@@ -7,32 +7,37 @@
         private const double fade_in_duration_multiplier = 0.4;
         private const double fade_out_duration_multiplier = 0.3;
+        private bool increaseFirstObjectVisibility = true;
+        private IEnumerable<DrawableHitObject> drawables;
```
again, should be unnecessary

smoogipoo replied:
I think we'll want this, but used in a different way. We'll want to bind to the bindable's valuechanged and apply/revert this - suppose the user changes the setting during the lead-in.

peppy replied:
interesting edge case. may want to handle at a higher level than per-mod.. not sure.

smoogipoo commented on osu.Game.Rulesets.Osu/Mods/OsuModHidden.cs:
```diff
@@ -7,32 +7,37 @@
+                foreach (var d in drawables.OfType<DrawableOsuHitObject>())
+                {
+                    //Don't hide the first object
+                    if (d.ChildID == 1 && increaseFirstObjectVisibility) continue;
```
We should not be using ChildID - consider using `drawables.First()` instead.

Vidalee committed:
Changes requests

Vidalee commented:
I changed the code as you requested (not really sure if I did correclty what peppy wanted though), however I can't pass a testcase, the TestCaseReplay one (osu.Game.Tests.Visual).

I have a timeout exception on it, I tried logging things etc, and it seems like for the osu!taiko, mania and catch, the SetReplay() void at line 290 (on my latest commit) is not executed (while the test is waiting for it).
I compared my "Chesterton's fence" commit which built correctly to the "Review changes" commit but found nothing that could explain the apparition of this bug.
I also tested many different cases, it appears that if I put `applyMods(Mods, null);` at the end of the RulesetContainer constructor everything works fine, but it doesn't make sense since (I tested) the applyMods() void is executed even in failed testcases. I have absolutely no clue why this error is spawning, do you have an idea?

Also for the 3 others hidden test case errors, they appeared because of the `d == drawables.Last()` that smoogipoo suggested (in the OsuModHidden file). Have I implemented it the wrong way, or is it something else? (still works with `d.childID == 1` by the way)

Aergwyn commented on osu.Game/Rulesets/UI/RulesetContainer.cs:
```diff
@@ -282,6 +290,9 @@ private void applyMods(IEnumerable<Mod> mods)
             foreach (var mod in mods.OfType<IApplicableToRulesetContainer<TObject>>())
                 mod.ApplyToRulesetContainer(this);
+
+            foreach (var mod in mods.OfType<IReadFromConfig>())
```
Shouldn't it read from config before applying to anything as it could change the outcome?

Aergwyn commented:
The tests seem to fail as the mod now tries to access a Bindable which is never set in the test case.

Vidalee committed:
makes more sense but not fixed

Vidalee commented:
Did it this way but on another local branch and pushed the wrong one  :thinking:. Anyway, it still don't work. I don't see why without touching anything related to the Bindable "increaseFirstObjectVisibility" it would suddently not work. (remember that if I change d == drawables.Last() to d.childID == 1, it works)

VPeruS commented:
it happens because now you need manually call ReadFromConfig() before using mod and currently only Ruleset container doing it.

VPeruS commented on osu.Game.Rulesets.Osu/Mods/OsuModHidden.cs:
```diff
@@ -7,24 +7,33 @@
-
         private const double fade_in_duration_multiplier = 0.4;
         private const double fade_out_duration_multiplier = 0.3;
+        private Bindable<bool> increaseFirstObjectVisibility;
```
you may create object and set to false by default

Aergwyn replied:
Can we change the name to `firstObjectVisibility` (same with the setting) as it's a "we show (or not)" question.

Vidalee committed:
Fixing hidden test bug

Vidalee committed:
build fix

Aergwyn commented on osu.Game/Rulesets/Mods/IReadFromConfig.cs:
```diff
@@ -0,0 +1,12 @@
+
+namespace osu.Game.Rulesets.Mods
+{
+    public interface IReadFromConfig
```
Can we get a little XML doc on this interface? A short one like all the others should be totally enough.

Vidalee committed:
added description of the IReadFromConfig interface

Vidalee committed:
cleared comment

Vidalee committed:
cleared comment

Vidalee commented:
Peppy suggested me to name the setting like this, and I understand him since we don't make the circle completely disapear, we add it's approach circle or not

Aergwyn commented on osu.Game.Rulesets.Osu/Mods/OsuModHidden.cs:
```diff
@@ -7,24 +7,35 @@
+                //Don't hide the first object if the increaseFirstObjectVisibility is true ("drawables" are in a reverse order -> Last() )
+                //If increaseFirstObjectVisibility is null then we're exectuing one of the tests for the hidden mod. Since it's too much
+                //work for just this interface to work on this tests (DI of the config..), I just make sure to not throw a null exception
+                if (d == drawables.Last() && (increaseFirstObjectVisibility == null || increaseFirstObjectVisibility)) continue;
```
Wouldn't initialising `increaseFirstObjectVisibility` fix all problems with test and make comments like above unnecessary?
`private Bindable<bool> increaseFirstObjectVisibility = new Bindable<bool>();`?

jorolf replied:
You could've used the "??" operator as well

Vidalee replied:
Totally forgot about this operator, thanks for the reminder :)

Vidalee committed:
increaseFirstObjectVisibility now cannot be null

ColdVolcano commented on osu.Game.Rulesets.Osu/Mods/OsuModHidden.cs:
```diff
@@ -7,24 +7,33 @@
             foreach (var d in drawables.OfType<DrawableOsuHitObject>())
             {
+                //Don't hide the first object if the increaseFirstObjectVisibility is true ("drawables" are in a reverse order -> Last() )
+                if (d == drawables.Last() && (increaseFirstObjectVisibility == null || increaseFirstObjectVisibility)) continue;
```
No need for null check now

VPeruS replied:
If you didn't check it will fail

Vidalee committed:
was persuaded I changed the if statement...

smoogipoo commented on osu.Game.Rulesets.Osu/Mods/OsuModHidden.cs:
```diff
@@ -7,24 +7,33 @@
 namespace osu.Game.Rulesets.Osu.Mods
 {
-    public class OsuModHidden : ModHidden, IApplicableToDrawableHitObjects
+    public class OsuModHidden : ModHidden, IApplicableToDrawableHitObjects, IReadFromConfig
```
This interface should be attached to the base `ModHidden.

smoogipoo commented on osu.Game.Rulesets.Osu/Mods/OsuModHidden.cs:
```diff
@@ -7,24 +7,33 @@
             foreach (var d in drawables.OfType<DrawableOsuHitObject>())
             {
+                //Don't hide the first object if the increaseFirstObjectVisibility is true ("drawables" are in a reverse order -> Last() )
+                if (d == drawables.Last() && increaseFirstObjectVisibility) continue;
```
We should probably be moving this to the base implementation. I.e.:

```
class ModHidden
{
    public void ApplyToDrawableHitObjects(IEnumerable<DrawableHitObject> drawables)
    {
        foreach (var d in drawables.OfType<DrawableHitObject>())
        {
             if (d == drawables.Last() && increaseHitObjectVisibility)
                 return;
             ApplyHiddenState(d)
        }
    }

    protected virtual void ApplyHiddenState(DrawableHitObject hitObject) { }
}

Vidalee replied:
The hidden mods for the other gamemodes do not need to hide (or not) the first hidden object, only the standard gamemode need it, so should we really implement it in a class (ModHidden) that all the hidden mods inherit?

peppy replied:
The idea was to bring all rulesets in-line with this behaviour, so @smoogipoo's approach would be correct here.

peppy committed:
Merge remote-tracking branch 'upstream/master' into Vidalee-osu-hd-setting

Vidalee committed:
Merge branch 'master' into osu-hd-setting

Vidalee committed:
Made the changes requested

Vidalee committed:
Code sanity

Vidalee committed:
White space trimmed

Vidalee committed:
Should be ok now

Vidalee committed:
Missed one white space.. I guess my editor wasn't properly configured
for this file

Vidalee committed:
Solving conflict

Vidalee committed:
Appvtests (#1)

* Resolving AppVeyor's errors

Vidalee committed:
Merge branch 'master' into osu-hd-setting

Vidalee committed:
resolving codefactor styling issues

peppy committed:
Merge branch 'master' into osu-hd-setting

peppy committed:
Reduce spacing changes

peppy committed:
Reorder methods and remove useless comments

peppy committed:
Reword settings text to be ruleset agnostic

peppy commented on osu.Game.Rulesets.Osu/Mods/OsuModHidden.cs:
```diff
@@ -2,38 +2,22 @@
-            {
-                d.ApplyCustomUpdateState += ApplyHiddenState;
-
-                d.HitObject.TimeFadein = d.HitObject.TimePreempt * fade_in_duration_multiplier;
```
what happened to this code..?

Vidalee replied:
Honestly no idea how I forgot to move that part too, and since everything was still working I didn't noted it :/

Vidalee committed:
why this line was ommited

Vidalee committed:
code sanity

Vidalee committed:
resolving the lost code problem

Vidalee committed:
resolving the lost code problem

peppy committed:
Merge branch 'master' into osu-hd-setting

peppy committed:
Move setting to new "mods" section

peppy committed:
Fix regressions

peppy committed:
Improve readability of code

peppy committed:
Formatting fixes
